### PR TITLE
feat(dns): add dns discovery service

### PR DESCRIPTION
**Pull request by mattsse**
WIP

adds DNS discovery implementation

## Update
* Service scaffold
* Parse support for all types

## TODO
- [ ] tree lookups
- [ ] service impl

@onbjerg opening this up for review now, can either finish this in this one or followup so it doesn't get too big, wdyt?

### CODE REVIEW

File: crates/net/dns/src/tree.rs [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+    #[error("Field {0} not found.")]
+    FieldNotFound(&'static str),
+    #[error("Base64 decoding failed: {0}")]
+    Base64DecodeError(String),
+    #[error("Base32 decoding failed: {0}")]
+    Base32DecodeError(String),
+    #[error("{0}")]
+    RlpDecodeError(String),
+    #[error("{0}")]
+    Other(String),
```
Comment (onbjerg) [inline crates/net/dns/src/tree.rs:68]: Either make these transparent or add context

File: crates/net/dns/src/tree.rs [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+            DnsEntry::Node(entry) => entry.fmt(f),
+        }
+    }
+}
+
+/// Error while parsing a [DnsEntry]
+#[derive(thiserror::Error, Debug)]
+#[allow(missing_docs)]
+pub enum ParseDnsEntryError {
+    #[error("Unknown Entry: {0}")]
```
Comment (onbjerg) [inline crates/net/dns/src/tree.rs:57]:
```suggestion
    #[error("Unknown entry: {0}")]
```

File: crates/net/dns/src/config.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#[derive(Debug, Clone)]
+pub struct DnsDiscoveryConfig {
+    /// Timeout for DNS lookups.
+    ///
+    /// Default: 5s
+    pub lookup_timeout: Duration,
+    /// The rate at which lookups should be re-triggered.
+    ///
+    /// Default: 30min
+    pub lookup_interval: Duration,
```
Comment (onbjerg) [inline crates/net/dns/src/config.rs:13]: Is this the lookup on the root node or to walk the entries in the tree?

**Review by onbjerg — APPROVED** [review]
I think it's ok to merge for now and finish it up in follow up PRs, added some comments for areas to clarify

File: crates/net/dns/src/tree.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/dns/src/tree.rs:68]: unfortunately the rlp::Error is not reexported from `enr` crate so I'd need to import the crate just for this error variant

File: crates/net/dns/src/config.rs (hunk 3 above)
Comment (mattsse) [reply crates/net/dns/src/config.rs:13]: this will be the interval to revisit trees

**Pull request merged**